mapping/react-leaflet.md

```diff
@@ -0,0 +1,433 @@
+        <>
+            <BaseMap center={[41.8781, -87.6298]} zoom={10}>
+            {/* this will only show when wardBorders has a value */}
+            {wardBorders && <GeoJSON
+                                key='ward-layer'
+                                data={wardBorders}
+                                style={fill} 
+                                onEachFeature={eventHandlersOnEachFeature} />}
+            </BaseMap>
+            {ward && <p>Ward {ward.ward}'s shape_area = {ward.shape_area} and shape_leng = {ward.shape_leng}</p>}
+        </>
+    )
+}
+
+export default ChicagoWardMap
+```
+
+Nice!
+
+### Lift state up
+One thing I don't like about the current design of the `ChicagoWardMap`: what if you need to use the ward's information elsewhere in the UI, in a place that is outside of this component? You would need to "lift state up" and set the `ward` state in a parent component. This is a common pattern in React applications, and can be difficult to understand if you're not familiar with it. It's simply a JavaScript callback that passes a message from a child component to a parent component.
+
+In your `App.js`, import the `useState` hook from React, create a `ward` state object, and render it on the page like you did in `ChicagoWardMap`:
+```jsx
+import { useState } from 'react'
+import './App.css'
+import ChicagoWardMap from './maps/wards'
+
+function App() {
+  const [ward, setWard] = useState(null)
+
+  return (
+    <div className="App">
+      <div className='map-viewer'>
+        <ChicagoWardMap 
+          onSelectWard={setWard}
+        />
+        {ward && <p>Ward {ward.ward}'s shape_area = {ward.shape_area} and shape_leng = {ward.shape_leng}</p>}
+      </div>
+    </div>
+  );
+}
+
+export default App
+```
+
+This code does what `ChicagoWardMap` did, but it depends on the `onSelectWard` prop to do this. So, we need to add that prop to the `ChicagoWardMap` component. (TODO: this is where a diff would be helpful...) Where you declare the `ChicagoWardMap` function, pass in `onSelectWard` as a prop:
+```jsx
+function ChicagoWardMap({ onSelectWard }) {...}
+```
+
+Remove these two lines of code, since you replaced them in `App.js`:
+- `const [ward, setWard] = useState({})`
+- `{ward && <p>Ward {ward.ward}'s shape_area = {ward.shape_area} and shape_leng = {ward.shape_leng}</p>}`
+
+And in the `onWardClick` function, replace the `setWard` function with `onChooseCounty`:
+```jsx
+function onCountyClick(e) {
+        const layer = e.target
+
+        const layerFeature = (layer && layer.feature && layer.feature.properties) 
+                                ? layer.feature.properties 
+                                : null
+
+        // send the information about the feature to the parent component
+        onChooseCounty(layerFeature)
+    }
+```
+
+If you did this correctly, your map should render the same as it did. This pattern can make applications complicated, so only use it sparingly, when it's absolutely necessary.
+
```

Comment:
At a high level, maps seem like a good time to use a class-based component because we have to define all of these helper methods. Do you have thoughts on class- vs. function-based components?

Reply:
@hancush This is a big debate! I'm partial to functional components. But to be open-minded, I reimplemented the `ChicagoWardMap` component as a class-based component. You can look at the difference here: 
- [Class](https://github.com/smcalilly/hello-leaflet/blob/main/src/maps/wards-class-component.js)
- [Functional](https://github.com/smcalilly/hello-leaflet/blob/main/src/maps/wards-functional-component.js)

Regarding class vs functional, the React documentation recommends using functional components over class-based components going forward, though they claim to never deprecate class-based components. [Their documentation gives some background about React Hooks]( https://reactjs.org/docs/hooks-intro.html#motivation), which is a big reason why one would want to use functional components. 

A functional component gives a developer a lot of extra functionality via the Hooks API, like `useContext`, `useDispatch`, `useMemo`, `userReducer`. You can see more about the API here: https://reactjs.org/docs/hooks-reference.html. (The FAQ in that section also delve into this discussion of class vs functional.) 

The Hooks API isn't available in a class-based component. Since the `react-leaflet` library leverages the Hooks API (as seen [in this example where you can access the map via the useMapEvent hook](https://react-leaflet.js.org/docs/example-events/)), I think functional components might be better for maps. Also, when I first started writing this documentation, I found that the latest versions of react-leaflet no longer support class-based components.

IMO, hooks are a more straightforward way of accessing React state (technically a functional component is "stateless" but it still manages the state in a functional programming way (from what i kinda understand about functional programming, still not certain about this)). I especially like how it's a descriptive, readable way of using state, since you essentially wrap the state in a name:
```jsx
// class
this.setState({
  wardBorders: features
)}

// functional
setWardBorders(feature)
```

In a class component, you have to use the various component lifecycle methods, like the confusing `componentDidMount`, `componentDidUpdate`, `componentWillUnmount`, rather the `useEffect` hook that always executes when the component renders, updates, and unmounts:
```jsx

// functional
useEffect(() => {
   const geojson = getGeoJSON() // psuedocode for the sake of example
   setWardBoders(geojson)
}, [setWardBorders]) // this "dependency array" is the second argument to the useEffect hook. it prevents unnecessary functional calls upon re-renders, and it cleans up the component after its lifecyle ends


// class
componentDidMount() {
    const geojson = getGeoJSON() // psuedocode for the sake of example
    this.setState({
      wardBorders: geojson
    })
}
  
componentWillUnmount() {
  // clean up the component however you need to 
}
```

I think that a functional component is the best way to write React components, but I also think this is something that might come down to a developer's preference and preferred style of coding, the team's preference, and whether or not the Hooks API would be useful for the use-case. Personally, at my first software dev job where I wrote a lot of React, some of the functional programming concepts found in React were the first more advanced concepts that clicked for me in understanding how to code. So I'm partial to functional React. It's only at DataMade where I've been doing more OOP. I think both paradigms are useful in certain contexts, but I'm not experienced or knowledgeable enough to know either way.

Reply:
This is a great discussion! Can you add a version of this to this documentation, with links to both examples since you have already implemented them?

Reply:
I also want to call out that, while I understand the sentiment, I very much disagree with "but I'm not experienced or knowledgeable enough to know either way." – especially when it comes at the end of a well-reasoned comment citing your experience and knowledge. 🙃